File: lib/transcode_response.dart
import 'package:spectrum_plugin/image/image_specfication.dart';

class TranscodeResponse {
  final bool isSuccessful;
  final String ruleName;
  final int totalBytesRead;
  final int totalBytesWritten;
  final ImageSpecification inputImageSpecification;
  final ImageSpecification outputImageSpecification;

  TranscodeResponse({
    this.isSuccessful,
    this.ruleName,
    this.totalBytesRead,
    this.totalBytesWritten,
    this.inputImageSpecification,
    this.outputImageSpecification,
  });

  @override
  String toString() {
    return "TranscodeResponse{" +
        "isSuccessful=" +
        isSuccessful.toString() +
        ", ruleName='" +
        (ruleName ?? "") +
        '\'' +
        ", inputImageSpecification=" +
        inputImageSpecification.toString() +
        ", outputImageSpecification=" +
        outputImageSpecification.toString() +
        ", totalBytesRead=" +
        totalBytesRead.toString() +
        ", totalBytesWritten=" +
        totalBytesWritten.toString() +
        '}';
  }

  static TranscodeResponse parse(Map<String, dynamic> map) {
    return TranscodeResponse(
      isSuccessful: map['isSuccessful'],
      ruleName: map['ruleName'],
      totalBytesRead: map['totalBytesRead'],
      totalBytesWritten: map['totalBytesWritten'],
      inputImageSpecification: ImageSpecification.parse(
        Map<String, dynamic>.from(map['inputImageSpecification'] ?? Map()),
      ),
      outputImageSpecification: ImageSpecification.parse(
        Map<String, dynamic>.from(map['outputImageSpecification'] ?? Map()),
      ),
    );
  }
}

File: android/src/main/kotlin/br/com/bugs/bunny/plugin/spectrum/TranscodeResponse.kt
package br.com.bugs.bunny.plugin.spectrum

import com.facebook.spectrum.SpectrumResult
import com.facebook.spectrum.image.ImageSize
import com.facebook.spectrum.image.ImageSpecification


class TranscodeResponse {
  object Builder {
    fun build(sr: SpectrumResult): Map<String, Any?> {
      return mapOf(
          "isSuccessful" to sr.isSuccessful,
          "ruleName" to sr.ruleName,
          "totalBytesRead" to sr.totalBytesRead,
          "totalBytesWritten" to sr.totalBytesWritten,
          "inputImageSpecification" to sr.inputImageSpecification?.asMap(),
          "outputImageSpecification" to sr.outputImageSpecification?.asMap()
      )
    }

    private fun ImageSpecification.asMap(): Map<String, Any?> {
      return mapOf(
          "chromaSamplingMode" to chromaSamplingMode?.name,
          "format" to format.identifier.toUpperCase(),
          "orientation" to orientation.name.toUpperCase(),
          // NOTE: There is no implementation available in the library for metadata :-/
          "metadata" to mapOf<String, Any>(),
          "size" to size.asMap(),
          "pixelSpecification" to pixelSpecification.name.toUpperCase()
      )
    }

    private fun ImageSize.asMap(): Map<String, Int> {
      return mapOf(
          "width" to width,
          "height" to height
      )
    }

  }
}